File: phi_plugin_koishi/model/fCompute.ts
import { MAX_DIFFICULTY } from './constNum'
import { logger } from '../components/Logger'
import { Session } from 'koishi'
import { h } from 'koishi'
import getInfo from './getInfo'
import { songString } from './type/type'

export default class compute {
    /**
     * 计算等效rks
     * @param {number} acc 
     * @param {number} difficulty 
     * @returns 
     */
    static rks(acc: number, difficulty: number) {
        if (acc == 100) {
            /**满分原曲定数即为有效rks */
            return Number(difficulty)
        } else if (acc < 70) {
            /**无效acc */
            return 0
        } else {
            /**非满分计算公式 [(((acc - 55) / 45) ^ 2) * 原曲定数] */
            return difficulty * (((acc - 55) / 45) * ((acc - 55) / 45))
        }
    }

    /**
     * 计算所需acc
     * @param {Number} rks 目标rks
     * @param {Number} difficulty 定数
     * @param {Number} [count=undefined] 保留位数
     * @returns 所需acc
     */
    static suggest(rks: number, difficulty: number, count: number = undefined): string | number {
        let ans = 45 * Math.sqrt(rks / difficulty) + 55

        if (ans >= 100)
            return "无法推分"
        else {
            if (count != undefined) {
                return `${ans.toFixed(count)}%`
            } else {
                return ans
            }
        }
    }

    /**
     * 发送文件
     * @param e 
     * @param file 
     * @param title 文件名称
     */
    static async sendFile(session: Session, file: Buffer, title: string = "") {
        try {
            session.send(h.file(file, "application/octet-stream", { title }))
        } catch (err) {
            logger.error(`文件上传错误：`)
            logger.error(err)
            await session.send(`文件上传错误` + err.message)
        }
    }

    /**
     * 获取角色介绍背景曲绘
     * @param {string} save_background 
     * @returns 
     */
    static async getBackground(save_background: string) {
        try {
            switch (save_background) {
                case 'Another Me ': {
                    save_background = 'Another Me (KALPA)'
                    break
                }
                case 'Another Me': {
                    save_background = 'Another Me (Rising Sun Traxx)'
                    break
                }
                case 'Re_Nascence (Psystyle Ver.) ': {
                    save_background = 'Re_Nascence (Psystyle Ver.)'
                    break
                }
                case 'Energy Synergy Matrix': {
                    save_background = 'ENERGY SYNERGY MATRIX'
                    break
                }
                case 'Le temps perdu-': {
                    save_background = 'Le temps perdu'
                    break
                }
                default: {
                    break
                }
            }
            return getInfo.getill(getInfo.SongGetId(save_background as songString))
        } catch (err) {
            logger.error(`获取背景曲绘错误`, err)
            return false
        }
    }

    /**
     * 为数字添加前导零
     * @param {number} num 原数字
     * @param {number} cover 总位数
     * @returns 前导零数字
     */
    static ped(num: number, cover: number) {
        return String("0".repeat(cover) + num).slice(-cover)
    }

    /**
     * 标准化分数
     * @param {number} score 分数
     * @returns 标准化的分数 0'000'000
     */
    static std_score(score: number) {
        let s1 = Math.floor(score / 1e6)
        let s2 = Math.floor(score / 1e3) % 1e3
        let s3 = score % 1e3
        return `${s1}'${this.ped(s2, 3)}'${this.ped(s3, 3)}`
    }

    /**
     * 随机数，包含上下界
     * @param {number} min 最小值
     * @param {number} max 最大值
     * @returns 随机数
     */
    static randBetween(min: number, max: number) {
        return Math.floor(Math.random() * (max - min + 1) + min)
    }

    /**
     * 随机打乱数组
     * @param {Array} arr 原数组
     * @returns 随机打乱的数组
     */
    static randArray(arr: Array<any>) {
        let newArr = []
        while (arr.length > 0) {
            newArr.push(arr.splice(Math.floor(Math.random() * arr.length), 1)[0])
        }
        return newArr
    }

    /**
     * 转换时间格式
     * @param {Date|string} date 时间
     * @returns 2020/10/8 10:08:08
     */
    static formatDate(date: Date | string) {
        date = new Date(date)
        return `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()} ${date.toString().match(/([0-9])+:([0-9])+:([0-9])+/)[0]}`
    }

    /**
     * 转换unity富文本
     * @param {string} richText 
     * @param {boolean} [onlyText=false] 是否只返回文本
     * @returns 
     */
    static convertRichText(richText: string, onlyText: boolean = false) {
        if (!richText) {
            return richText
        }
        richText = richText.replace(/</g, '&lt;').replace(/>/g, '&gt;');
        let reg = [/&lt;color\s*=\s*.*?&gt;(.*?)&lt;\/color&gt;/, /&lt;size\s*=\s*.*?&gt;(.*?)&lt;\/size&gt;/, /&lt;i&gt;(.*?)&lt;\/i&gt;/, /&lt;b&gt;(.*?)&lt;\/b&gt;/]
        while (1) {
            if (richText.match(reg[0])) {
                let txt = richText.match(reg[0])[1]
                let color = richText.match(reg[0])[0].match(/&lt;color\s*=\s*(.*?)&gt;/)[1].replace(/[\s\"]/g, '')
                richText = richText.replace(reg[0], onlyText ? txt : `<span style="color:${color}">${txt}</span>`)
                continue
            }

            if (richText.match(reg[2])) {
                let txt = richText.match(reg[2])[1]
                richText = richText.replace(reg[2], onlyText ? txt : `<i>${txt}</i>`)
                continue
            }

            if (richText.match(reg[3])) {
                let txt = richText.match(reg[3])[1]
                richText = richText.replace(reg[3], onlyText ? txt : `<b>${txt}</b>`)
                continue
            }
            // if (richText.match(reg[1])) {
            //     let txt = richText.match(reg[1])[1]
            //     let size = richText.match(reg[1])[0].match(/size\s*=[^>]*?([^>]*)/)[1]o
            //     return this.convertRichText(richText.replace(reg[1], `<span style="font-size:${size}px">${txt}</span>`))
            // }
            if (richText.match(/\n\r?/)) {
                richText.replace(/\n\r?/g, '<br>')
            }
            break
        }
        if (onlyText) {
            richText = richText.replace(/&lt;/g, '<').replace(/&gt;/g, '>');
        }
        return richText
    }

    /**是否是管理员 */
    static authority(session: Session) {
        let sessionRoles = session?.event?.member?.roles
        return sessionRoles && (sessionRoles.includes('admin') || sessionRoles.includes('owner'))
    }

    /**
     * 捕获消息中的范围
     * @param {string} msg 消息字符串
     * @param {Array} range 范围数组
     * @param {number} [max=MAX_DIFFICULTY] 最大值
     * @param {number} [min=0] 最小值
     */
    static match_range(msg: string, range: Array<any>, max: number = MAX_DIFFICULTY, min: number = 0) {
        range[0] = min
        range[1] = max
        if (msg.match(/[0-9]+(.[0-9]+)?\s*[-～~]\s*[0-9]+(.[0-9]+)?/g)) {
            /**0-16.9 */
            msg = msg.match(/[0-9]+(.[0-9]+)?\s*[-～~]\s*[0-9]+(.[0-9]+)?/g)[0]
            let result = msg.split(/\s*[-～~]\s*/g)
            range[0] = Number(result[0])
            range[1] = Number(result[1])
            if (range[0] > range[1]) {
                let tem = range[1]
                range[1] = range[0]
                range[0] = tem
            }
            if (range[1] % 1 == 0 && !result.includes(".0")) range[1] += 0.9
        } else if (msg.match(/[0-9]+(.[0-9]+)?\s*[-+]/g)) {
            /**16.9- 15+ */
            msg = msg.match(/[0-9]+(.[0-9]+)?\s*[-+]/g)[0]
            let result = msg.replace(/\s*[-+]/g, '')
            if (msg.includes('+')) {
                range[0] = Number(result)
            } else {
                range[1] = Number(result)
                if (range[1] % 1 == 0 && !result.includes(".0")) range[1] += 0.9
            }
        } else if (msg.match(/[0-9]+(.[0-9]+)?/g)) {
            /**15 */
            msg = msg.match(/[0-9]+(.[0-9]+)?/g)[0]
            range[0] = range[1] = Number(msg)
            if (!msg.includes('.')) {
                range[1] += 0.9
            }
        }

    }

    /**
     * 匹配消息中对成绩的筛选
     * @param {string} msg 
     * @param {number} [max=MAX_DIFFICULTY] 范围最大值
     * @param {number} [min=0] 范围最小值
     * @returns 
     */
    static match_request(msg: string = "", max: number = MAX_DIFFICULTY, min: number = 0) {
        let range = [min, max]

        /**EZ HD IN AT */
        let isask = [true, true, true, true]

        msg = msg.toUpperCase()

        if (msg.includes('EZ') || msg.includes('HD') || msg.includes('IN') || msg.includes('AT')) {
            isask = [false, false, false, false]
            if (msg.includes('EZ')) { isask[0] = true }
            if (msg.includes('HD')) { isask[1] = true }
            if (msg.includes('IN')) { isask[2] = true }
            if (msg.includes('AT')) { isask[3] = true }
        }
        msg = msg.replace(/(list|AT|IN|HD|EZ)*/g, "")

        let scoreAsk = { NEW: true, F: true, C: true, B: true, A: true, S: true, V: true, FC: true, PHI: true }

        if (msg.includes(' NEW') || msg.includes(' F') || msg.includes(' C') || msg.includes(' B') || msg.includes(' A') || msg.includes(' S') || msg.includes(' V') || msg.includes(' FC') || msg.includes(' PHI')) {
            scoreAsk = { NEW: false, F: false, C: false, B: false, A: false, S: false, V: false, FC: false, PHI: false }
            let rating = ['NEW', 'F', 'C', 'B', 'A', 'S', 'V', 'FC', 'PHI']
            for (let i in rating) {
                if (msg.includes(` ${rating[i]}`)) { scoreAsk[rating[i]] = true }
            }
        }
        if (msg.includes(` AP`)) { scoreAsk.PHI = true }
        msg = msg.replace(/(NEW|F|C|B|A|S|V|FC|PHI|AP)*/g, "")

        this.match_range(msg, range, max, min)
        return { range, isask, scoreAsk }
    }

    static async recallMsg(session: Session, messageId: string) {
        return await session.bot.deleteMessage(session.channelId, messageId)
    }

    static async sleep(ms: number) {
        return new Promise(resolve => setTimeout(resolve, ms))
    }


    /**
     * 转换时间格式
     * @param {Date|string} date 时间
     * @returns 2020/10/8 10:08:08
     */
    static date_to_string(date: Date | string | number): string {
        if (!date) return undefined
        date = new Date(date)

        let month = (date.getMonth() + 1) < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1
        let day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()

        return `${date.getFullYear()}/${month}/${day} ${date.toString().match(/([0-9])+:([0-9])+:([0-9])+/)[0]}`
    }


    // 定义一个函数，不接受参数，返回一个随机的背景色
    static getRandomBgColor() {
        // 生成三个 0 到 200 之间的随机整数，分别代表红、绿、蓝分量
        let red = Math.floor(Math.random() * 201);
        let green = Math.floor(Math.random() * 201);
        let blue = Math.floor(Math.random() * 201);
        // 将三个分量转换为十六进制形式，然后拼接成一个 RGB 颜色代码
        let hexColor = "#" + this.toHex(red) + this.toHex(green) + this.toHex(blue);
        // 返回生成的颜色代码
        return hexColor;
    }

    // 定义一个函数，接受一个整数参数，返回它的十六进制形式
    static toHex(num: number) {
        // 如果数字小于 16，就在前面补一个 0
        if (num < 16) {
            return "0" + num.toString(16);
        } else {
            return num.toString(16);
        }
    }


    /**
     * 
     * @param {number} real_score 真实成绩
     * @param {number} tot_score 总成绩
     * @param {boolean} fc 是否fc
     * @returns 
     */
    static rate(real_score: number, tot_score: number, fc: boolean) {

        if (!real_score) {
            return 'F'
        } else if (real_score == tot_score) {
            return 'phi'
        } else if (fc) {
            return 'FC'
        } else if (real_score >= tot_score * 0.96) {
            return 'V'
        } else if (real_score >= tot_score * 0.92) {
            return 'S'
        } else if (real_score >= tot_score * 0.88) {
            return 'A'
        } else if (real_score >= tot_score * 0.82) {
            return 'B'
        } else if (real_score >= tot_score * 0.70) {
            return 'C'
        } else {
            return 'F'
        }
    }

    /**
     * 计算百分比
     * @param {Number} value 值
     * @param {Array} range 区间数组 (0,..,1)
     * @returns 百分数，单位%
     */
    static percentage(value: number, range: number[]): number {
        if (range[0] == range[range.length - 1]) {
            return 50
        } else {
            return (value - range[0]) / (range[range.length - 1] - range[0]) * 100
        }
    }

}

File: phi_plugin/model/fCompute.py
from datetime import datetime
import math
import random
import re

from nonebot.adapters import Bot
from nonebot_plugin_alconna import AlconnaMatcher, File
from nonebot_plugin_uninfo import Uninfo

from zhenxun.models.level_user import LevelUser
from zhenxun.utils.rules import ensure_group
from zhenxun.utils.withdraw_manage import WithdrawManager

from ..components.Logger import logger
from ..utils import Date
from .constNum import MAX_DIFFICULTY


class fCompute:
    @staticmethod
    def rks(acc: float, difficulty: float):
        """计算等效rks"""
        if acc == 100:
            # 满分原曲定数即为有效rks
            return difficulty
        elif acc < 70:
            # 无效acc
            return 0
        else:
            # 非满分计算公式 [(((acc - 55) / 45) ^ 2) * 原曲定数]
            return difficulty * (((acc - 55) / 45) ** 2)

    @staticmethod
    def suggest(rks: float, difficulty: float, count: int | None = None) -> str | float:
        """
        计算所需acc

        :param rks: 目标rks
        :param difficulty: 定数
        :param count: 保留位数
        """
        ans = 45 * math.sqrt(rks / difficulty) + 55
        if ans >= 100:
            return "无法推分"
        elif count is not None:
            return f"{ans:.{count}f}%"
        else:
            return ans

    @staticmethod
    async def sendFile(matcher: AlconnaMatcher, file: bytes, title: str = "backup.zip"):
        """发送文件"""
        try:
            await matcher.send(File(raw=file, name=title))
        except Exception as e:
            logger.error("文件上传错误", "phi-plugin", e=e)
            await matcher.send(f"文件上传错误.{type(e)}: {e}")

    @staticmethod
    async def getBackground(save_background: str) -> str | None:
        """获取角色介绍背景曲绘"""
        try:
            match save_background:
                case "Another Me ":
                    save_background = "Another Me (KALPA)"
                case "Another Me":
                    save_background = "Another Me (Rising Sun Traxx)"
                case "Re_Nascence (Psystyle Ver.) ":
                    save_background = "Re_Nascence (Psystyle Ver.)"
                case "Energy Synergy Matrix":
                    save_background = "ENERGY SYNERGY MATRIX"
                case "Le temps perdu-":
                    save_background = "Le temps perdu"
            return save_background
        except Exception as e:
            logger.error("获取背景曲绘错误", "phi-plugin", e=e)
            return None

    @staticmethod
    def ped(num: float, cover: int) -> str:
        """
        为数字添加前导零

        :param num: 原数字
        :param cover: 总位数
        """
        return str(int(num)).zfill(cover)

    @staticmethod
    def std_score(score: float) -> str:
        """
        标准化分数

        :param score: 分数
        :return: 标准化的分数 0'000'000
        """
        s1 = math.floor(score / 1e6)
        s2 = math.floor(score / 1e3) % 1e3
        s3 = score % 1e3
        return f"{s1}'{fCompute.ped(s2, 3)}'{fCompute.ped(s3, 3)}"

    @staticmethod
    def randBetween(min: int, max: int) -> int:
        """随机数，包含上下界"""
        return random.randint(min, max)

    @staticmethod
    def randArray(arr: list) -> list:
        """随机打乱数组"""
        return random.sample(arr, len(arr))

    @staticmethod
    def formatDate(date: datetime | str) -> str:
        """
        转换时间格式

        :return: 2020/10/8 10:08:08
        """
        date = Date(date)
        return date.strftime("%Y/%m/%d %H:%M:%S")

    @staticmethod
    def convertRichText(richText: str, onlyText: bool = False) -> str:
        """
        转换unity富文本

        :param richText: richText
        :param onlyText: 是否只返回文本
        """
        if not richText:
            return richText
        while True:
            matched = False

            # 处理颜色标签
            color_match = re.search(r"<color\s*=\s*.*?>(.*?)</color>", richText)
            if color_match:
                full_match = color_match[0]
                txt = color_match[1]

                if color_attr := re.search(r"<color\s*=\s*(.*?)>", full_match):
                    color = re.sub(r'[\s"]', "", color_attr[1])
                    replacement = (
                        txt if richText else f'<span style="color:{color}">{txt}</span>'
                    )
                    richText = re.sub(
                        r"<color\s*=\s*.*?>(.*?)</color>",
                        replacement,
                        richText,
                        count=1,
                    )
                    matched = True
                    continue

            if italic_match := re.search(r"<i>(.*?)</i>", richText):
                txt = italic_match[1]
                replacement = txt if onlyText else f"<i>{txt}</i>"
                richText = re.sub(r"<i>(.*?)</i>", replacement, richText, count=1)
                matched = True
                continue

            if bold_match := re.search(r"<b>(.*?)</b>", richText):
                txt = bold_match[1]
                replacement = txt if onlyText else f"<b>{txt}</b>"
                richText = re.sub(r"<b>(.*?)</b>", replacement, richText, count=1)
                matched = True
                continue

            if "\n" in richText or "\r" in richText:
                richText = re.sub(r"\r?\n", "<br>", richText)
                matched = True

            if not matched:
                break

        return richText

    @staticmethod
    async def authority(session: Uninfo) -> bool:
        """是否是管理员"""
        groupId = session.scene.id if ensure_group(session) else None
        return await LevelUser.get_user_level(session.user.id, groupId) >= 5

    @staticmethod
    def match_range(
        msg: str, range_list: list, max_val: float = MAX_DIFFICULTY, min_val: float = 0
    ) -> list[float]:
        """
        捕获消息中的数值范围

        :param msg: 消息字符串
        :param range_list: 要修改的列表
        :param max_val: 最大值
        :param min_val: 最小值
        """
        range_list[0] = min_val
        range_list[1] = max_val
        if result := re.search(r"(\d+\.?\d*)\s*[-～~]\s*(\d+\.?\d*)", msg):
            # 0-16.9
            start_str, end_str = result.groups()
            start, end = float(start_str), float(end_str)

            range_list[0] = start
            range_list[1] = end

            # 保证范围顺序正确
            if start > end:
                range_list[0], range_list[1] = end, start

            # 特殊处理：当上限为整数且原始字符串不含小数点时
            if float(range_list[1]).is_integer() and "." not in end_str:
                range_list[1] += 0.9
        elif result := re.search(r"(\d+\.?\d*)\s*([-+])", msg):
            # 16.9- 15+
            num_str, symbol = result.groups()
            num = float(num_str)

            if symbol == "+":
                range_list[0] = num
            else:  # symbol == '-'
                range_list[1] = num
                # 特殊处理：当数值为整数且原始字符串不含小数点时
                if num.is_integer() and "." not in num_str:
                    range_list[1] += 0.9
        elif result := re.search(r"(\d+\.?\d*)", msg):
            num_str = result[1]
            num = float(num_str)

            range_list[0] = range_list[1] = num

            # 特殊处理：当原始字符串不含小数点时
            if "." not in num_str:
                range_list[1] += 0.9
        return range_list

    @staticmethod
    def match_request(
        msg: str, max_val: float = MAX_DIFFICULTY, min_val: float = 0
    ) -> dict:
        """
        匹配消息中对成绩的筛选

        :param msg: 消息内容
        :param max: 范围最大值
        :param min: 范围最小值
        """
        range_list = [min_val, max_val]
        # EZ HD IN AT
        isask = [True, True, True, True]
        msg = msg.upper()
        if "EZ" in msg or "HD" in msg or "IN" in msg or "AT" in msg:
            isask = [False, False, False, False]
        if "EZ" in msg:
            isask[0] = True
        if "HD" in msg:
            isask[1] = True
        if "IN" in msg:
            isask[2] = True
        if "AT" in msg:
            isask[3] = True
        msg = (
            msg.replace("EZ", "").replace("HD", "").replace("IN", "").replace("AT", "")
        )
        scoreAsk = {
            "NEW": True,
            "F": True,
            "C": True,
            "B": True,
            "A": True,
            "S": True,
            "V": True,
            "FC": True,
            "PHI": True,
        }
        if (
            "NEW" in msg
            or "F" in msg
            or "C" in msg
            or "B" in msg
            or "A" in msg
            or "S" in msg
            or "V" in msg
            or "FC" in msg
            or "PHI" in msg
        ):
            scoreAsk = {
                "NEW": False,
                "F": False,
                "C": False,
                "B": False,
                "A": False,
                "S": False,
                "V": False,
                "FC": False,
                "PHI": False,
            }
        for i in scoreAsk:
            if i in msg:
                scoreAsk[i] = True
        if " AP" in msg:
            scoreAsk["PHI"] = True
        for i in scoreAsk:
            msg = msg.replace(i, "")
        range_list = fCompute.match_range(msg, range_list, max_val, min_val)
        return {"range": range_list, "scoreAsk": scoreAsk, "isask": isask}

    @staticmethod
    async def recallMsg(bot: Bot, messageId: str):
        await WithdrawManager.withdraw_message(bot, messageId)

    @staticmethod
    def date2str(date: datetime | str | float | None) -> str | None:
        """
        转换时间格式

        :return: 2020/10/08 10:08:08
        """
        return Date(date).strftime("%Y/%m/%d %H:%M:%S") if date else None

    @staticmethod
    def getRandomBgColor() -> str:
        """定义一个函数，不接受参数，返回一个随机的背景色"""
        red = random.randint(0, 200)
        green = random.randint(0, 200)
        blue = random.randint(0, 200)
        return f"#{fCompute.toHex(red)}{fCompute.toHex(green)}{fCompute.toHex(blue)}"

    @staticmethod
    def toHex(num: int) -> str:
        """定义一个函数，接受一个整数参数，返回它的十六进制形式"""
        return hex(num)[2:]

    @staticmethod
    def rate(real_score: float, tot_score: float, fc: bool) -> str:
        """
        :param real_score: 真实成绩
        :param tot_score: 总成绩
        :param fc: 是否fc
        """
        if not real_score:
            return "F"
        elif real_score >= tot_score:
            return "phi"
        elif fc:
            return "FC"
        elif real_score >= tot_score * 0.96:
            return "V"
        elif real_score >= tot_score * 0.92:
            return "S"
        elif real_score >= tot_score * 0.88:
            return "A"
        elif real_score >= tot_score * 0.82:
            return "B"
        elif real_score >= tot_score * 0.70:
            return "C"
        else:
            return "F"

    @staticmethod
    def percentage(value: float, range_list: list[float]) -> float:
        """
        计算百分比

        :param value: 值
        :param range_list: 区间数组 (0,..,1)
        :return: 百分数，单位%
        """
        if range_list[0] == range_list[-1]:
            return 50
        else:
            return round(
                ((value - range_list[0]) / (range_list[-1] - range_list[0]) * 100),
                2,
            )
